### DISCUSSION TITLE
How do I prevent the settings menu and download window from opening in the background?

### DISCUSSION THREAD

**Discussion by patrickdobler** [asmvik/yabai#2139] [Q&A]
Hello, I have a very annoying issue that I have been trying to find a solution for a long time.

When I open a settings menu of an application or a download window pops up in Firefox, it briefly appears in the foreground for a few milliseconds and then disappears into the background. I have to use Alt+Tab (or select in the dock) to bring the window back to the front / into focus.

I guess it's a macOS or Yabai setting causing this. The problem occurs with applications like bsp as well as with stack layout.

Setup:
- 2 external monitors with multiple spaces
- Sonoma 14.3.1
- Yabai 6.0.15
- mouse_follows_focus and focus_follows_mouse = off
- Except for application rules, space config for stack or float, and more or less standard config parameters, nothing is configured in the yabairc. There are no other "focus" options set.

Example:
![windows_background](https://github.com/koekeishiya/yabai/assets/16536946/69b40e5a-c5da-4584-97ef-80fd3cb4aa8d)

**Comment by tjex** [comment]
can you post your yabai config?

Sonoma has a bug whereby newly launched applications don't always take focus, and quitting an application will often focus Finder instead of the last focussed app. It could be related to that, but this behaviour looks weird. As at least with the Sonoma bug, the window should still be visible, albeit not focussed.

**Reply by patrickdobler** [reply]
Hi, thanks for looking into it. I still have Sonoma 14.3.1.

I also corrected a typo in my first message. "The problem occurs with applications like bsp as well as with stack layout."

```
#!/usr/bin/env sh

# Disable automatic tiling
yabai -m rule --add app="^Calculator$" manage=off
yabai -m rule --add app="^Karabiner-Elements$" manage=off
yabai -m rule --add app="^Karabiner-EventViewer$" manage=off
yabai -m rule --add app="^Xnapper$" manage=off
yabai -m rule --add app="^Shottr$" manage=off
#yabai -m rule --add app="^Finder$" manage=off
yabai -m rule --add app="^Keka$" manage=off
yabai -m rule --add app="^Disk Utility$" manage=off
yabai -m rule --add app="^Activity Monitor$" manage=off
yabai -m rule --add app="^Terminal$" manage=off
yabai -m rule --add app="^AltTab$" manage=off
yabai -m rule --add app="^Raycast$" manage=off
yabai -m rule --add app="^Logitech G HUB$" manage=off
yabai -m rule --add app="^LinearMouse$" manage=off
yabai -m rule --add app="^Dropover$" manage=off
yabai -m rule --add app="^Pure Paste$" manage=off
yabai -m rule --add app="^Contacts$" manage=off
yabai -m rule --add app="^qView$" manage=off
yabai -m rule --add app="^Clock$" manage=off
yabai -m rule --add app="^Infuse$" manage=off
yabai -m rule --add app="^IINA$" manage=off
yabai -m rule --add app="^Parallels Desktop$" manage=off
yabai -m rule --add app="^Bartender 5$" manage=off

# float system preferences
yabai -m rule --add app="^System Information$" manage=off
yabai -m rule --add app="^System Settings$" title=".*" manage=off
yabai -m rule --add title='Preferences$' manage=off
yabai -m rule --add title='Settings$' manage=off
yabai -m rule --add title='^Archive Utility$' manage=off

# set some popup window/settings menu to float
yabai -m rule --add label="Finder" app="^Finder$" title="(Co(py|nnect)|Move|Info|Pref)" manage=off
yabai -m rule --add label="Safari" app="^Safari$" title="^(General|(Tab|Password|Website|Extension)s|AutoFill|Se(arch|curity)|Privacy|Advance|Advanced)$" manage=off
yabai -m rule --add label="Microsoft Excel" app="^Microsoft Excel$" title="Microsoft Visual Basic" manage=off
yabai -m rule --add label="TradingView" app="^TradingView$" title="(TradingView|Login )" manage=off
yabai -m rule --add label="Firefox" app="^Firefox$" title="Opening " manage=off
yabai -m rule --add label="Parallels Desktop" app="^Parallels Desktop$" title="Control Center " manage=off

# custom space window
yabai -m rule --add app="^TradingView$" space=1
yabai -m rule --add app="^Remote Desktop Manager$" space=4
yabai -m rule --add app="^(Firefox|Safari)$" space=3
#yabai -m rule --add app="^(Microsoft Excel)$" space=4
yabai -m rule --add app="^(Microsoft Edge|Parallels Desktop)$" space=5
yabai -m rule --add app="^(WhatsApp|Discord|Proton Mail|Fantastical)$" space=6

yabai -m space 6 --layout stack
yabai -m space 5 --layout float

CENTER_APPS="^(Bartender 5|1Password)$"
yabai -m rule --add app="${CENTER_APPS}" manage=off grid=4:4:1:1:2:2


# global settings
yabai -m config mouse_follows_focus          off                # center mouse on window with focus
yabai -m config focus_follows_mouse          off                # [autofocus|autoraise|off] https://apple.stackexchange.com/questions/269622/macos-windows-requiring-an-explicit-click-to-make-active-before-ui-elements-ins
yabai -m config window_origin_display        cursor             # specify which display a newly created window should be managed in. default: The display in which the window is created (standard macOS behaviour). focused: The display that has focus when the window is created. cursor: The display that currently holds the mouse cursor.
yabai -m config window_placement             second_child       # New window spawns to the right if vertical split, or bottom if horizontal split
yabai -m config window_zoom_persist          on            
yabai -m config window_topmost               off                # floating windows are always on top, required SIP disabled (default: off)
yabai -m config window_shadow                on            
yabai -m config window_animation_duration    0.0           
yabai -m config window_animation_frame_rate  120           
yabai -m config window_opacity_duration      0.0           
yabai -m config active_window_opacity        1.0           
yabai -m config normal_window_opacity        0.90          
yabai -m config window_opacity               off           
yabai -m config insert_feedback_color        0xffd75f5f    
yabai -m config active_window_border_color   0xfffcba03
yabai -m config normal_window_border_color   0x00FFFFFF    
yabai -m config window_border_width          1
yabai -m config window_border_radius         12            
yabai -m config window_border_blur           off           
yabai -m config window_border_hidpi          on            
yabai -m config window_border                off            
yabai -m config split_ratio                  0.50          
yabai -m config split_type                   auto          
yabai -m config auto_balance                 off           
yabai -m config top_padding                  8             
yabai -m config bottom_padding               8             
yabai -m config left_padding                 8             
yabai -m config right_padding                8             
yabai -m config window_gap                   8            
yabai -m config layout                       bsp              # default layout (can be bsp, stack or float)
yabai -m config mouse_modifier               fn               # modifier for clicking and dragging with mouse
yabai -m config mouse_action1                move             # set modifier + left-click drag to move window
yabai -m config mouse_action2                resize           # set modifier + right-click drag to resize window
yabai -m config mouse_drop_action            swap             # when window is dropped in center of another window, swap them (on edges it will split it)
yabai -m config external_bar                 off:40:0
yabai -m config menubar_opacity              1.0              # https://github.com/koekeishiya/yabai/issues/2091
```

**Reply by tjex** [reply]
Firstly, you're missing the line to load the scripting addition. I think it's required, and not optional? That could be creating some weird behaviour? The shebang also isnt needed from my understanding?

eg:
```
yabai -m signal --add event=dock_did_restart action="sudo yabai --load-sa"
sudo yabai --load-sa
```
[see here in the docs](https://github.com/koekeishiya/yabai/wiki/Installing-yabai-(latest-release)#configure-scripting-addition)

Otherwise, what happens if you uncomment the rule for Fantastical?

`yabai -m rule --add app="^(WhatsApp|Discord|Proton Mail|Fantastical)$" space=6`

Perhaps, yabai is applying the rule to the settings window as well, and either sending it to space 6 (in the case where fantastical wasn't on space 6 already), or 'sending it' to the current space (space 6) but it ends up behind the other window?..

**Comment by patrickdobler** [comment]
Thank you for your suggestions. I think I don't need the first one because I use yabai without disabling System Integrity Protection, as I don't need the additional features.

The behavior affected every application, not just Fantastical, and even those without any specific rules. However, you made a good point that it could be due to different space rules.

I've just finished setting up my new desk and made some changes to the config file for space allocation. I no longer keep my MacBook's lid open and only use the BSP layout. Now, everything works fine. Here are the changes I made:

Old Setup:
* MacBook: Space 6 with stack layout
* Display 1: Spaces 1 and 2
* Display 2: Spaces 3 and 4 with BSP layout, and Space 5 with float layout

New Setup:
* Four external displays, each with two spaces, all using BSP layout.

I'll work with my new setup for 1-2 days and if it still works I'll close this answer, although it's probably not fixed just because I changed my setup.

Thanks again @tjex for your time and help, I really appreciate it!

**Reply by tjex** [reply]
" I no longer keep my MacBook's lid open", this could have been the problem too. There is a setting that needs to be enabled in MacOS for yabai compatibility, something like "allow per monitor spaces".

**Reply by patrickdobler** [reply]
Do you mean one of these?
- Automatically rearrange Spaces based on most recent use: OFF
- Displays have separate Spaces: ON

**Reply by tjex** [reply]
yep, both of those options. 
Describe here for future visitors: https://github.com/koekeishiya/yabai/wiki#installation-requirements

**Comment by tjex** [comment]
@patrickdobler maybe this is of use (setting a layer rule): https://github.com/koekeishiya/yabai/issues/1929#issuecomment-1872581528

**Comment by patrickdobler** [comment]
Out of curiosity, I tried to test it with a similar setup as before and got the same problem with popups opening in the background again. It probably has something to do with my Macbook being open and being the main display. I will close this question now as I have changed my setup and everything is working as expected so far (even for applications where the window is not in the rule configured space).
